controllers/spannerautoscaleschedule_controller.go

```diff
@@ -86,29 +92,64 @@ func (r *SpannerAutoscaleScheduleReconciler) Reconcile(ctx context.Context, req
 		return ctrlreconcile.Result{}, nil
 	}
 
+	sasFinalizerName := "spannerautoscaleschedule.spanner.mercari.com/finalizer"
+
+	// register our finalizer if the object is not being deleted
+	if sas.ObjectMeta.DeletionTimestamp.IsZero() && !ctrlutil.ContainsFinalizer(&sas, sasFinalizerName) {
+		log.V(1).Info("adding finalizer to schedule", "schedule", sas, "finalizer", sasFinalizerName)
+		ctrlutil.AddFinalizer(&sas, sasFinalizerName)
+		if err := r.ctrlClient.Update(ctx, &sas); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
 	var sa spannerv1beta1.SpannerAutoscaler
 	nnsa := types.NamespacedName{
 		Name:      sas.Spec.TargetResource,
-		Namespace: nn.Namespace,
+		Namespace: req.NamespacedName.Namespace,
 	}
-	if err := r.ctrlClient.Get(ctx, nnsa, &sa); err != nil {
-		err = ctrlclient.IgnoreNotFound(err)
-		if err != nil {
-			log.Error(err, "failed to get spanner-autoscaler")
-			return ctrlreconcile.Result{}, err
+	err := r.ctrlClient.Get(ctx, nnsa, &sa)
+	if err != nil && sas.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Do not ignore NotFound error, we need a valid SpannerAutoscaler
+		log.Error(err, "failed to get spanner-autoscaler")
+		r.recorder.Event(&sas, corev1.EventTypeWarning, "SpannerAutoscalerNotFound", err.Error())
+
+		return ctrlreconcile.Result{}, err
+	}
+
+	// if the object is being deleted, perform clean up
+	if !sas.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is being deleted
+		if ctrlutil.ContainsFinalizer(&sas, sasFinalizerName) {
+			// our finalizer is present, so if we found any autoscaler resource, then remove the schedule from the autoscaler list
+			if len(sa.Status.Schedules) != 0 {
+				sa.Status.Schedules = deleteFromArray(sa.Status.Schedules, req.NamespacedName.Name)
+				log.Info("deleting schedule from spanner-autoscaler", "schedule", sas.Name, "autoscaler", sa.Name)
+				if err := r.ctrlClient.Status().Update(ctx, &sa); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+
+			// remove our finalizer from the list and update it.
+			ctrlutil.RemoveFinalizer(&sas, sasFinalizerName)
+			if err := r.ctrlClient.Update(ctx, &sas); err != nil {
+				return ctrl.Result{}, err
+			}
 		}
 
-		return ctrlreconcile.Result{}, nil
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
 	}
 
 	log.Info("registering schedule with spanner-autoscaler", "schedule", sas.Name, "autoscaler", sa.Name)
 
-	if _, ok := findInArray(sa.Status.Schedules, nn.Name); !ok {
-		sa.Status.Schedules = append(sa.Status.Schedules, nn.Name)
+	if _, ok := findInArray(sa.Status.Schedules, req.NamespacedName.Name); !ok {
+		sa.Status.Schedules = append(sa.Status.Schedules, req.NamespacedName.Name)
 		if err := r.ctrlClient.Status().Update(ctx, &sa); err != nil {
 			return ctrl.Result{}, err
```

Comment:
It seems better to output the log.
Like this https://github.com/mercari/spanner-autoscaler/blob/master/controllers/spannerautoscaler_controller.go#L284-L285

Reply:
Thanks, fixed!